sys/dev/drm/drmkpi/drmcompat_dma_buf.c

```diff
@@ -0,0 +1,425 @@
+
+	attach = malloc(sizeof(*attach), M_DMABUF, M_WAITOK | M_ZERO);
+	attach->dmabuf = dmabuf;
+	attach->dev = dev;
+
+	sx_xlock(&dmabuf->db_sx);
+	if (dmabuf->ops->attach != NULL)
+		rv = dmabuf->ops->attach(dmabuf, attach);
+	sx_xunlock(&dmabuf->db_sx);
+	if (rv != 0) {
+		free(attach, M_DMABUF);
+		return ERR_PTR(rv);
+	}
+
+	return (attach);
+}
+
+void
+dma_buf_detach(struct dma_buf *dmabuf, struct dma_buf_attachment *attach)
+{
+	MPASS(dmabuf != NULL);
+	MPASS(attach != NULL);
+
+	if (dmabuf == NULL || attach == NULL)
+		return;
+
+	sx_xlock(&dmabuf->db_sx);
+	if (dmabuf->ops->detach)
+		dmabuf->ops->detach(dmabuf, attach);
+	sx_xunlock(&dmabuf->db_sx);
+
+	free(attach, M_DMABUF);
+}
+
+struct sg_table *
+dma_buf_map_attachment(struct dma_buf_attachment *attach,
+    enum dma_data_direction dir)
+{
+
+	MPASS(attach != NULL);
+	MPASS(attach->dmabuf != NULL);
+
+	if (attach == NULL || attach->dmabuf == NULL)
+		return (ERR_PTR(-EINVAL));
+
+	return (attach->dmabuf->ops->map_dma_buf(attach, dir));
+}
+
+void
+dma_buf_unmap_attachment(struct dma_buf_attachment *attach,
+    struct sg_table *sg, enum dma_data_direction dir)
+{
+
+	MPASS(attach != NULL);
+	MPASS(attach->dmabuf != NULL);
+
+	if (attach == NULL || attach->dmabuf == NULL)
+		return;
+
+	return (attach->dmabuf->ops->unmap_dma_buf(attach, sg, dir));
+}
+
+static int
+dmabuf_fop_close(struct file *file, struct thread *td)
+{
+	struct dma_buf *dmabuf;
+
+	if (!file_is_dmabuf(file))
+		return (EINVAL);
+
+	dmabuf = file->f_data;
+
+	dmabuf->ops->release(dmabuf);
+	reservation_poll_fini(&dmabuf->db_resv_poll);
+
+	if (dmabuf->resv == &dmabuf->db_resv_int[0])
+		reservation_object_fini(dmabuf->resv);
+
+	sx_destroy(&dmabuf->db_sx);
+	free(dmabuf, M_DMABUF);
+	return (0);
+}
+
+static int
+dmabuf_fop_poll(struct file *file, int events, struct ucred *active_cred,
+    struct thread *td)
+{
+	struct dma_buf *dmabuf;
+	struct reservation_poll *rpoll;
+
+	if (!file_is_dmabuf(file))
+		return (EINVAL);
+
+	dmabuf = file->f_data;
+
+	rpoll = &dmabuf->db_resv_poll;
+
+	return (reservation_object_poll(dmabuf->resv, events, rpoll));
+}
+
+static int
+dmabuf_fop_kqfilter(struct file *file, struct knote *kn)
+{
+	struct dma_buf *dmabuf;
+	struct reservation_poll *rpoll;
+
+	if (!file_is_dmabuf(file))
+		return (EINVAL);
+
+	dmabuf = file->f_data;
+
+	rpoll = &dmabuf->db_resv_poll;
+
+	return (reservation_object_kqfilter(dmabuf->resv, kn, rpoll));
+}
+
+static int
+dmabuf_fop_mmap(struct file *file, vm_map_t map, vm_pointer_t *addr,
+	    vm_offset_t max_addr, vm_size_t size, vm_prot_t prot,
+	    vm_prot_t cap_maxprot, int flags, vm_ooffset_t foff,
+	    struct thread *td)
+{
+	struct dma_buf *dmabuf;
+	struct vm_area_struct vma;
+
+	if (!file_is_dmabuf(file))
+		return (EINVAL);
+
+	dmabuf = file->f_data;
+
+	if (foff + size  > dmabuf->size)
+		return (EINVAL);
+
+	if (*addr + size > max_addr)
+		return (EINVAL);
+
+	vma.vm_start = *addr;
+	vma.vm_end = *addr + size;
+	vma.vm_pgoff = foff;
```

Comment:
Consumers seems to assume other fields are initialized. 
```suggestion
	vma = {
		.vm_start = *addr;
		.vm_end = *addr + size;
		.vm_pgoff = foff;
	};
```

Reply:
The fact that vm_flags isn't set suggests to me that this function is never used and doesn't work.